### PR TITLE
Add an gRPC endpoint for request verification

### PR DESCRIPTION
**Pull request by ntchjb**
Fixed: BAND-220

When new oracle request is arrived on bandchain, yoda received the event and trying to request data from data provider. When there is a data provider that requires payment in order to access the data, the data provider should setup a server to check the request sent by yoda before providing data. The server may use this gRPC endpoint to verify the request sent by yoda to make sure that data provider will provide data to correct reporter account.

## Implementation details

The gRPC endpoint is at `/oracle/verify_request`. It requires the following information for verification before report the data

- Chain ID
- A validator address
- A request ID
- An data source's external ID
- Public key of a reporter
- Report's signature for the content of ` RequestVerification` proto message.

It will verify the following logic

- Request should not be empty
- Provided chain ID should match validator's chain ID
- Provided signature should be valid, which means this query request should be signed by the given reporter
- Provided reporter should be authorized by the given validator
- Provided request ID should exist on chain
- Provided validator should be assigned to response to the request
- Provided external ID should be required by the request determined by oracle script
- Provided validator should not have reported data for the request
- The request should not be expired

There is also CLI command provided for ease of use and testing.

```sh
bandd query oracle verify-request [chain-id] [validator-addr] [request-id] [data-source-external-id] [reporter-addr] [reporter-signature-hex] [flags]
```

Please note that in unit test, there is a method that will be deprecated and unsupported soon, begins at v0.43.0, which is `Bech32ifyPubKey`  (also affeects `MustBech32ifyPubKey`). Start at v0.43.0, we need to change the way to input reporter's public key by encoding the public key using protobuf's `Any` type field. See https://github.com/cosmos/cosmos-sdk/issues/7357 for more information

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: x/oracle/keeper/grpc_query.go [hunk 1]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	// Request should not be empty
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	// Provided chain ID should match validator's chain ID
+	if ctx.ChainID() != req.ChainId {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("provided chain ID does not match the validator's chain ID; expected %s, got %s", ctx.ChainID(), req.ChainId))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:170]: codes.BadRequest (400) on this case should be better.

does not match the validator's chain => does not match the current chain ID

File: x/oracle/keeper/grpc_query.go [hunk 2]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	reporter := sdk.AccAddress(reporterPubKey.Address().Bytes())
+	isReporterAuthorizedByValidator := false
+	for _, existingReporter := range reporters {
+		if reporter.Equals(existingReporter) {
+			isReporterAuthorizedByValidator = true
+			break
+		}
+	}
+	if !isReporterAuthorizedByValidator {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("%s is not an authorized reporter of %s", reporter, req.Validator))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:201]: codes.UnAuthorize?

File: x/oracle/keeper/grpc_query.go [hunk 3]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	}
+
+	// Provided request should exist on chain
+	request, err := k.GetRequest(ctx, types.RequestID(req.RequestId))
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("unable to get request from chain: %s", err.Error()))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:207]: codes.NotFound

File: x/oracle/keeper/grpc_query.go [hunk 4]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	isValidatorAssigned := false
+	for _, requestedValidator := range request.RequestedValidators {
+		v, _ := sdk.ValAddressFromBech32(requestedValidator)
+		if validator.Equals(v) {
+			isValidatorAssigned = true
+			break
+		}
+	}
+	if !isValidatorAssigned {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("%s is not assigned for request ID %d", validator, req.RequestId))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:220]: codes.UnAuthorize?

File: x/oracle/keeper/grpc_query.go [hunk 5]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	// Provided external ID should be required by the request determined by oracle script
+	var dataSourceID *types.DataSourceID
+	for _, rawRequest := range request.RawRequests {
+		if rawRequest.ExternalID == types.ExternalID(req.ExternalId) {
+			dataSourceID = &rawRequest.DataSourceID
+			break
+		}
+	}
+	if dataSourceID == nil {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("no data source required by the request %d found which relates to the external data source with ID %d.", req.RequestId, req.ExternalId))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:232]: I think it should return 502 (Internal server error). If a request existed it should have a data source for every raw requests

File: x/oracle/keeper/grpc_query.go [hunk 6]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	isValidatorReported := false
+	for _, report := range reports {
+		reportVal, _ := sdk.ValAddressFromBech32(report.Validator)
+		if reportVal.Equals(validator) {
+			isValidatorReported = true
+			break
+		}
+	}
+	if isValidatorReported {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("validator %s already submitted data report for this request", validator))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:246]: Let's think which error code suit this case.

File: x/oracle/keeper/grpc_query.go [hunk 7]
```diff
@@ -155,3 +156,107 @@ func (k Querier) RequestSearch(c context.Context, req *types.QueryRequestSearchR
...
+	}
+
+	// The request should not be expired
+	params := k.GetParams(ctx)
+	if request.RequestHeight+int64(params.ExpirationBlockCount) < ctx.BlockHeader().Height {
+		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("Request with ID %d is already expired", req.RequestId))
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:252]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

File: x/oracle/keeper/grpc_query.go (hunk 1 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go:170]: There is no Bad Request error for gRPC status code. However, there is a mapping here:

https://grpc.github.io/grpc/core/md_doc_http-grpc-status-mapping.html

Shall we use `codes.INTERNAL` instead as an mapping of `Bad Request`?

File: x/oracle/keeper/grpc_query.go (hunk 5 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go:232]: I think there is another possibility that a client sends some random external ID that does not exists in the request, which causes data source ID to be nil.

File: x/oracle/keeper/grpc_query.go (hunk 6 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go:246]: Can it be `AlreadyExists`?

File: x/oracle/keeper/grpc_query.go (hunk 7 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go:252]: Can it be `DeadlineExceeded`?

File: x/oracle/keeper/grpc_query.go (hunk 2 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go]: There is no `Unauthorized`, but there is `Unauthenticated` instead, as it is mapping of each other.

File: proto/oracle/v1/query.proto [hunk 8]
```diff
@@ -69,6 +69,13 @@ service Query {
       returns (QueryRequestPriceResponse) {
     option (google.api.http).post = "/oracle/request_prices";
   }
+
+  // RequestVerification verifies a request to make sure that
+  // all information that will be used to report the data is valid
+  rpc RequestVerification(QueryRequestVerificationRequest)
+    returns (QueryRequestVerificationResponse) {
+      option (google.api.http).post = "/oracle/verify_request";
```
Comment (Benzbeeb) [inline proto/oracle/v1/query.proto:77]: please change to '/oracle/v1/verify_request'

**Review by Benzbeeb — CHANGES_REQUESTED** [review]

File: proto/oracle/v1/query.proto (hunk 8 above)
Comment (ntchjb) [reply proto/oracle/v1/query.proto]: Fix. Test passed, yay.

**Review by Benzbeeb — APPROVED** [review]

File: x/oracle/keeper/grpc_query.go (hunk 1 above)
Comment (taobun) [reply x/oracle/keeper/grpc_query.go]: I think this link help us more to use code https://pkg.go.dev/google.golang.org/grpc@v1.36.1/codes#Code

Comment (taobun) [reply x/oracle/keeper/grpc_query.go]: But not internal for sure. FailedPrecondition?

**Review by taobun — CHANGES_REQUESTED** [review]

File: x/oracle/keeper/grpc_query.go (hunk 1 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go]: Sounds good. Will use FailedPrecondition.

File: x/oracle/client/cli/query.go [hunk 9]
```diff
@@ -284,6 +287,51 @@ func GetQueryActiveValidators() *cobra.Command {
 // 	}
 // }
 
+// GetQueryRequestVerification implements the query request verification command.
+func GetQueryRequestVerification() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:  "verify-request [chain-id] [validator-addr] [request-id] [data-source-external-id] [reporter-addr] [reporter-signature-hex]",
```
Comment (taobun) [inline x/oracle/client/cli/query.go:293]: reporter-pubkey?

File: x/oracle/keeper/grpc_query.go [hunk 10]
```diff
@@ -156,6 +157,111 @@ func (k Querier) RequestPrice(c context.Context, req *types.QueryRequestPriceReq
...
+			isValidatorReported = true
+			break
+		}
+	}
+	if isValidatorReported {
+		return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("validator %s already submitted data report for this request", validator))
+	}
+
+	// The request should not be expired
+	params := k.GetParams(ctx)
```
Comment (taobun) [inline x/oracle/keeper/grpc_query.go:251]: It's time to use ExpirationBlockCount(ctx)

**Review by taobun — CHANGES_REQUESTED** [review]

File: x/oracle/client/cli/query.go (hunk 9 above)
Comment (ntchjb) [reply x/oracle/client/cli/query.go]: Yes, fixed.

File: x/oracle/keeper/grpc_query.go (hunk 10 above)
Comment (ntchjb) [reply x/oracle/keeper/grpc_query.go]: Fixed

**Review by taobun — APPROVED** [review]

**Pull request merged**